lib/puppet/functions/vault_lookup/lookup.rb

```diff
@@ -2,60 +2,238 @@
+        role_id = ENV[env_role_id]
+        role_token = get_role_auth_token(connection, vault_url, vault_namespace, vault_app_role_path, role_id, secret_id)
+        vault_token = role_token
+      elsif File.exist?(role_file)
+        role_id = File.open(role_file).read
+        role_token = get_role_auth_token(connection, vault_url, vault_namespace, vault_app_role_path, role_id.chomp, secret_id)
+        vault_token = role_token
+      else
+        raise Puppet::Error, "Role file #{role_file} required for role_id does not exist"
+      end
+
+    end
 
-    token = get_auth_token(connection)
+    secret_response = if vault_namespace.nil? || vault_namespace == ''
+                        connection.get(
+                          URI(vault_url + '/v1/' + path.to_s),
+                          options: { 'include_system_store' => true },
+                          headers: { 'X-Vault-Token' => vault_token },
+                        )
+                      else
+                        connection.get(
+                          URI(vault_url + '/v1/' + path.to_s),
+                          options: { 'include_system_store' => true },
+                          headers: { 'X-Vault-Token' => vault_token, 'X-Vault-Namespace' => vault_namespace },
+                        )
+                      end
 
-    secret_response = connection.get("/v1/#{path}", 'X-Vault-Token' => token)
-    unless secret_response.is_a?(Net::HTTPOK)
+    unless secret_response.success?
       message = "Received #{secret_response.code} response code from vault at #{uri.host} for secret lookup"
       raise Puppet::Error, append_api_errors(message, secret_response)
     end
 
     begin
-      data = JSON.parse(secret_response.body)['data']
+      jsondata = if path.include? '/data/'
+                   JSON.parse(secret_response.body)['data']['data']
+                 else
+                   JSON.parse(secret_response.body)['data']
+                 end
+      data = if key_field.nil? || key_field == ''
+               jsondata
+             else
+               jsondata[key_field]
+             end
     rescue StandardError
       raise Puppet::Error, 'Error parsing json secret data from vault response'
     end
 
     Puppet::Pops::Types::PSensitiveType::Sensitive.new(data)
   end
 
-  private
+private
+
+  def get_cert_auth_token(connection, vault_url, vault_namespace, vault_cert_path, vault_cert_role)
+    role_data = if vault_cert_role.nil? || vault_cert_role == ''
+                  ''
+                else
+                  "{\"name\": \"#{vault_cert_role}\"}"
+                end
+
+    response = if vault_namespace.nil? || vault_namespace == ''
+                 connection.post(
+                   URI(vault_url + '/v1/auth/' + vault_cert_path.to_s + '/login'),
+                   role_data,
+                   headers: { 'Content-Type' => 'application/json' },
+                   options: { 'include_system_store' => true },
```

Comment:
Same here:

```suggestion
                   options: { include_system_store: => true },
```

Reply:
@joshcooper when changed to `options: { include_system_store: => true }`  works fine with `options: { 'include_system_store' => true },`
```
Error: Failed to apply catalog: Received 400 response code from vault at https://vault.host:8200 for authentication (api errors: ["1 error occurred:\n\t* error running lookahead function for mfa: no client certificate found\n\n"])
```

Reply:
Using a symbol is correct. The client cert issue is https://tickets.puppetlabs.com/browse/PUP-11522.